pkg/scheduler/plugins/multihierarchyelasticquota/quota_info.go

```diff
@@ -0,0 +1,184 @@
+		},
+	}
+}
+
+func (qi *QuotaInfo) DeepCopy() *QuotaInfo {
+	qi.lock.Lock()
+	defer qi.lock.Unlock()
+
+	return &QuotaInfo{
+		Name:                             qi.Name,
+		ParentName:                       qi.ParentName,
+		IsParent:                         qi.IsParent,
+		AllowLentResource:                qi.AllowLentResource,
+		Type:                             qi.Type,
+		RuntimeVersion:                   qi.RuntimeVersion,
+		NeedScaleMinQuotaWhenOverRootRes: qi.NeedScaleMinQuotaWhenOverRootRes,
+		CalculateInfo: QuotaCalculateInfo{
+			Max:                       qi.CalculateInfo.Max.DeepCopy(),
+			CalMin:                    qi.CalculateInfo.CalMin.DeepCopy(),
+			OriMin:                    qi.CalculateInfo.OriMin.DeepCopy(),
+			Used:                      qi.CalculateInfo.Used.DeepCopy(),
+			Request:                   qi.CalculateInfo.Request.DeepCopy(),
+			SharedRatio:               qi.CalculateInfo.SharedRatio.DeepCopy(),
+			Runtime:                   qi.CalculateInfo.Runtime.DeepCopy(),
+			SubGroupsTotalResLastTime: qi.CalculateInfo.SubGroupsTotalResLastTime.DeepCopy(),
+		},
+	}
+}
+
+func (qi *QuotaInfo) GetLimitRequest() v1.ResourceList {
+	limitRequest := util.NewResourceList()
+	for resName, quantity := range qi.CalculateInfo.Request {
+		if maxQuantity, ok := qi.CalculateInfo.Max[resName]; ok {
+			if quantity.Cmp(maxQuantity) == -1 {
+				//req < Max
+				limitRequest[resName] = quantity.DeepCopy()
+			} else {
+				limitRequest[resName] = maxQuantity.DeepCopy()
+			}
+		}
+	}
+	return limitRequest
+}
+
+func (qi *QuotaInfo) GetIsParent() bool {
+	qi.lock.Lock()
+	defer qi.lock.Unlock()
+	return qi.IsParent
+}
+
+func (qi *QuotaInfo) UpdateDeltaRequest(delta v1.ResourceList) {
+	qi.CalculateInfo.Request = v12.Add(qi.CalculateInfo.Request, delta)
+}
+
+func (qi *QuotaInfo) UpdateDeltaUsed(delta v1.ResourceList) {
+	qi.CalculateInfo.Request = v12.Add(qi.CalculateInfo.Used, delta)
+}
+
+func (qi *QuotaInfo) SetMaxQuota(res v1.ResourceList) {
+	qi.CalculateInfo.Max = res.DeepCopy()
+}
+
+func (qi *QuotaInfo) SetOriMinQuota(res v1.ResourceList) {
+	qi.CalculateInfo.OriMin = res.DeepCopy()
+}
+
+func (qi *QuotaInfo) SetCalMinQuota(res v1.ResourceList) {
+	qi.CalculateInfo.CalMin = res.DeepCopy()
+}
+
+func (qi *QuotaInfo) SetSharedRatio(res v1.ResourceList) {
+	qi.CalculateInfo.SharedRatio = res.DeepCopy()
+}
+
+func (qi *QuotaInfo) LockGetRequest() v1.ResourceList {
+	qi.lock.Lock()
+	defer qi.lock.Unlock()
+	return qi.CalculateInfo.Request.DeepCopy()
+}
+
+func (qi *QuotaInfo) LockGetUsed() v1.ResourceList {
+	qi.lock.Lock()
+	defer qi.lock.Unlock()
+	return qi.CalculateInfo.Used.DeepCopy()
+}
+
+func (qi *QuotaInfo) LockGetOriMin() v1.ResourceList {
+	qi.lock.Lock()
+	defer qi.lock.Unlock()
+	return qi.CalculateInfo.OriMin.DeepCopy()
+}
+
+func NewQuotaInfoFromQuota(quota *v1alpha1.ElasticQuota) (*QuotaInfo, error) {
+	isParent, parentName, err := GetParentQuotaInfo(quota)
```

Comment:
I think in NewQuotaInfo should not return so many err; may be introduce a validQuotaCheck before new is better.